File: src/harness/agent/wizard/stagnation.py
"""
Stagnation detection for identifying and handling stuck execution.
Detects retry loops, identical outputs, and global stagnation.
"""

import hashlib
from dataclasses import dataclass
from typing import Dict, List, Optional, TYPE_CHECKING

if TYPE_CHECKING:
    from .work_ledger import WorkLedger
    from .worker import WorkerOutcome
    from .plan_state import PlanState
    from .plan_patch import PlanPatch


@dataclass
class StagnationSignal:
    """Signal indicating stagnation detected."""

    detected: bool
    severity: float  # 0.0 to 1.0
    reason: str
    step_num: Optional[int] = None
    suggested_action: str = ""


class StagnationDetector:
    """
    Detects execution stagnation for escalation.

    SIGNALS:
    1. Too many retries on same step
    2. Identical outputs (spinning)
    3. No progress across multiple steps
    """

    def __init__(
        self,
        max_retries_per_step: int = 3,
        max_identical_outputs: int = 3,
        no_progress_threshold: int = 10,
    ):
        self.max_retries_per_step = max_retries_per_step
        self.max_identical_outputs = max_identical_outputs
        self.no_progress_threshold = no_progress_threshold
        self._output_hashes: Dict[int, List[str]] = {}

    def check(
        self, step_num: int, ledger: "WorkLedger", outcome: Optional["WorkerOutcome"] = None
    ) -> StagnationSignal:
        """
        Check for stagnation signals.
        Returns signal if stagnation detected.
        """
        from .work_ledger import EntryStatus

        history = ledger.get_step_history(step_num)
        # Count only COMPLETED or FAILED attempts, not DISPATCHED (in-flight)
        # This prevents off-by-one where the current attempt is counted
        completed_attempts = sum(
            1 for entry in history
            if entry.status in (EntryStatus.COMPLETED, EntryStatus.FAILED)
        )
        if completed_attempts > self.max_retries_per_step:
            return StagnationSignal(
                detected=True,
                severity=0.8,
                reason=f"Step {step_num} failed {completed_attempts} times (max {self.max_retries_per_step})",
                step_num=step_num,
                suggested_action="skip_step",
            )

        if outcome and outcome.final_response:
            # MD5 used for output fingerprinting to detect stagnation, not security
            output_hash = hashlib.md5(outcome.final_response.encode(), usedforsecurity=False).hexdigest()[:8]
            self._output_hashes.setdefault(step_num, []).append(output_hash)
            hashes = self._output_hashes[step_num]
            if len(hashes) >= self.max_identical_outputs:
                recent = hashes[-self.max_identical_outputs :]
                if len(set(recent)) == 1:
                    return StagnationSignal(
                        detected=True,
                        severity=0.9,
                        reason=f"Step {step_num} producing identical outputs",
                        step_num=step_num,
                        suggested_action="pivot_approach",
                    )

        recent = ledger.get_recent_entries(self.no_progress_threshold)
        if len(recent) >= self.no_progress_threshold:
            completed = sum(1 for entry in recent if entry.status == EntryStatus.COMPLETED)
            if completed == 0:
                return StagnationSignal(
                    detected=True,
                    severity=1.0,
                    reason="No steps completed in recent work items",
                    suggested_action="abort_or_simplify",
                )

        return StagnationSignal(detected=False, severity=0.0, reason="")

    def escalate(
        self, signal: StagnationSignal, plan_state: "PlanState"
    ) -> Optional["PlanPatch"]:
        """
        Generate escalation patch based on stagnation signal.
        Returns patch to skip step or add recovery step.
        """
        from .plan_patch import PlanPatch

        if not signal.detected:
            return None

        if signal.suggested_action == "skip_step" and signal.step_num:
            return PlanPatch.create_remove(
                base_version=plan_state.version,
                step_num=signal.step_num,
                justification=f"Stagnation: {signal.reason}",
            )

        return None

    def reset_step(self, step_num: int) -> None:
        """Reset tracking for a step (called on successful completion or skip)."""
        self._output_hashes.pop(step_num, None)

    def cleanup_all(self) -> None:
        """Clear all tracking state (called when orchestration completes)."""
        self._output_hashes.clear()

File: src/agent-ts/wizard/stagnation.ts
/**
 * Stagnation detection for identifying and handling stuck execution.
 * Detects retry loops, identical outputs, and global stagnation.
 *
 * Ported from: src/harness/agent/wizard/stagnation.py
 */

import { createHash } from 'crypto';
import type { WorkLedger } from './work-ledger.js';
import { EntryStatus } from './work-ledger.js';
import type { WorkerOutcome } from './worker.js';
import type { PlanState } from './plan-state.js';

/**
 * Signal indicating stagnation detected.
 */
export interface StagnationSignal {
  detected: boolean;
  severity: number; // 0.0 to 1.0
  reason: string;
  stepNum?: number;
  suggestedAction: string;
}

/**
 * Create a non-stagnation signal.
 */
export function noStagnation(): StagnationSignal {
  return {
    detected: false,
    severity: 0,
    reason: '',
    suggestedAction: '',
  };
}

/**
 * Detects execution stagnation for escalation.
 *
 * SIGNALS:
 * 1. Too many retries on same step
 * 2. Identical outputs (spinning)
 * 3. No progress across multiple steps
 */
export class StagnationDetector {
  private maxRetriesPerStep: number;
  private maxIdenticalOutputs: number;
  private noProgressThreshold: number;
  private outputHashes = new Map<number, string[]>();

  constructor(
    maxRetriesPerStep = 3,
    maxIdenticalOutputs = 3,
    noProgressThreshold = 10
  ) {
    this.maxRetriesPerStep = maxRetriesPerStep;
    this.maxIdenticalOutputs = maxIdenticalOutputs;
    this.noProgressThreshold = noProgressThreshold;
  }

  /**
   * Check for stagnation signals.
   * Evaluates all conditions and returns the signal with highest severity.
   */
  check(stepNum: number, ledger: WorkLedger, outcome?: WorkerOutcome): StagnationSignal {
    const signals: StagnationSignal[] = [];

    const history = ledger.getStepHistory(stepNum);

    // Count only COMPLETED or FAILED attempts, not DISPATCHED (in-flight)
    const completedAttempts = history.filter(
      (entry) =>
        entry.status === EntryStatus.COMPLETED || entry.status === EntryStatus.FAILED
    ).length;

    if (completedAttempts > this.maxRetriesPerStep) {
      signals.push({
        detected: true,
        severity: 0.8,
        reason: `Step ${stepNum} failed ${completedAttempts} times (max ${this.maxRetriesPerStep})`,
        stepNum,
        suggestedAction: 'skip_step',
      });
    }

    // Check for identical outputs (spinning)
    if (outcome?.finalResponse) {
      const outputHash = createHash('md5')
        .update(outcome.finalResponse)
        .digest('hex')
        .slice(0, 8);

      const hashes = this.outputHashes.get(stepNum) ?? [];
      hashes.push(outputHash);
      this.outputHashes.set(stepNum, hashes);

      if (hashes.length >= this.maxIdenticalOutputs) {
        const recent = hashes.slice(-this.maxIdenticalOutputs);
        const uniqueHashes = new Set(recent);
        if (uniqueHashes.size === 1) {
          signals.push({
            detected: true,
            severity: 0.9,
            reason: `Step ${stepNum} producing identical outputs`,
            stepNum,
            suggestedAction: 'pivot_approach',
          });
        }
      }
    }

    // Check for global stagnation
    const recent = ledger.getRecentEntries(this.noProgressThreshold);
    if (recent.length >= this.noProgressThreshold) {
      const completed = recent.filter((e) => e.status === EntryStatus.COMPLETED).length;
      if (completed === 0) {
        signals.push({
          detected: true,
          severity: 1.0,
          reason: 'No steps completed in recent work items',
          suggestedAction: 'abort_or_simplify',
        });
      }
    }

    // Return the signal with highest severity, or noStagnation if none detected
    if (signals.length === 0) {
      return noStagnation();
    }

    return signals.reduce((highest, current) =>
      current.severity > highest.severity ? current : highest
    );
  }

  /**
   * Generate escalation action based on stagnation signal.
   */
  getEscalationAction(
    signal: StagnationSignal,
    _planState: PlanState
  ): { action: 'skip' | 'abort' | 'none'; stepNum?: number } {
    if (!signal.detected) {
      return { action: 'none' };
    }

    if (signal.suggestedAction === 'skip_step' && signal.stepNum !== undefined) {
      return { action: 'skip', stepNum: signal.stepNum };
    }

    if (signal.suggestedAction === 'abort_or_simplify') {
      return { action: 'abort' };
    }

    return { action: 'none' };
  }

  /**
   * Reset tracking for a step (called on successful completion or skip).
   */
  resetStep(stepNum: number): void {
    this.outputHashes.delete(stepNum);
  }

  /**
   * Clear all tracking state (called when orchestration completes).
   */
  cleanupAll(): void {
    this.outputHashes.clear();
  }
}
